core-bundle/src/Messenger/WebWorker.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Messenger;
+
+use Psr\Cache\CacheItemInterface;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Console\Input\ArrayInput;
+use Symfony\Component\Console\Output\NullOutput;
+use Symfony\Component\EventDispatcher\Attribute\AsEventListener;
+use Symfony\Component\HttpKernel\Event\TerminateEvent;
+use Symfony\Component\Messenger\Command\ConsumeMessagesCommand;
+use Symfony\Component\Messenger\Event\WorkerRunningEvent;
+use Symfony\Component\Messenger\Event\WorkerStartedEvent;
+
+/**
+ * This service accepts an array of Symfony Messenger transports and automatically
+ * detects which of those have real workers running. In case they do not, it falls
+ * back to a web worker (kernel.terminate event) to ensure, messages are always
+ * processed, no matter whether a real worker is running or not.
+ *
+ * Detecting works by using the WorkerStartedEvent and WorkerRunningEvent which
+ * sets/updates a cache items and allowing a grace period. The grace period defines how
+ * long after the last event the web worker should consider a real worker to be running.
+ * Imagine you only have one worker running, and it works on a message for 10 minutes.
+ * In such a case, no WorkerStartedEvent or WorkerRunningEvent will be triggered for 10
+ * minutes so if there were a grace period of only a few seconds, the web worker would
+ * jump in probably too quickly.
+ *
+ * In any case, this provides advantages as it allows us to always use a queue and
+ * postpone processes to at least kernel.terminate in case there are no real
+ * workers available!
+ */
+class WebWorker
+{
+    private bool $webWorkerRunning = false;
+
+    public function __construct(
+        private readonly CacheItemPoolInterface $cache,
+        private readonly ConsumeMessagesCommand $consumeMessagesCommand,
+        private readonly array $transports,
+        private readonly string $gracePeriod = 'PT10M',
+    ) {
+    }
+
+    #[AsEventListener(priority: -512)]
+    public function onKernelTerminate(TerminateEvent $event): void
```

Comment:
```suggestion
    public function onKernelTerminate(): void
```

The argument is not used.

Reply:
You cannot do that because Symfony needs to know which event to listen too.

Reply:
The alternative (if we really want to) would be

```php
#[AsEventListener(TerminateEvent::class, priority: -512)]
public function onKernelTerminate(): void
```

Personally I would not prefer that though.

Reply:
I know but we don't do that anywhere else, do we?

Reply:
No, we don‘t. I have already marked this conversation as resolved.